XIPs/xip-31-message-backup-request-system.md

```diff
@@ -0,0 +1,466 @@
+
+#### Flow for Message Backup Providers
+
+1. Receive a `MessageHistoryBackupRequest` as part of normal message receiving flow
+1. Retrieve the contact for the installation that sent the message, and ensure that it has not been revoked. If revoked, ignore.
+1. Ensure that the installation that sent the message has a contact signed by the same blockchain account as the current user. If not, ignore.
+1. Convert all messages in the local database into a Message Backup File (maybe we want to chunk here?)
+1. Generate ephemeral encryption key, salt, and nonce. Encrypt the file using these keys
+1. Upload the file to the `backupStorageProviderUploadUrl` from the request
+1. Reply to the message with a `MessageHistoryBackupResponse` containing the encryption details, the hash of the backup file, and the `backupUrl` provided by the Backup Storage Provider
+1. Delete the `MessageHistoryBackupRequest`, the local database dump, and all the encryption keys.
+
+#### End-to-end flow
+
+```mermaid
+sequenceDiagram
+    Participant B as Backup Requester
+    Participant S as Backup Storage Provider
+    Participant N as XMTP Network
+    Participant M as Message Backup Provider
+    B->>S: Request a backup storage upload URL
+    S-->>B: Receive one time upload URL
+    B->>N: Send a MessageHistoryBackupRequest to each potential Message Backup Provider
+    M->>N: Check for MessageHistoryBackupRequest
+    N-->>M: Receive MessageHistoryBackupRequest
+    M->>S: Upload backup file
+    M->>N: Send MessageHistoryBackupResponse
+    B->>N: Check for MessageHistoryBackupResponse
+    N-->>B: Receive response
+    B->>S: Request file from URL in response
+    S-->>B: Receive file
+    B-->>B: Decrypt and load file into database
+```
+
+### Backup Account Files
+
+Backup account files are designed as a solution for users who have lost access to all of their XMTP applications. To create a Backup Account File, an application will generate a new set of XMTP keys and register the installation on the XMTP network. The installation contact published to the network will be set with a type of `backup`, which instructs all well-behaved clients to disregard any messages from this installation other than `MessageHistoryBackupResponse`s. This means that a compromised backup installation cannot be used to send messages on the user's behalf, which makes these files a less valuable target for attack.
+
+Once a backup account's contact has been published to the network, their installation messaging topic will receive a copy of all new messages sent to the associated blockchain account.
+
+The newly generated XMTP keys are encrypted and downloaded to the user's device. They may be encrypted by either a passphrase or a wallet signature. Users or applications may decide to store the Backup Account File in a private cloud storage provider (iCloud, Google Drive, etc). Or the file could simply be stored on-device with the risk of loss if the device was inaccessible.
```

Comment:
> The minimal friction case is storing the backup on the same device as the app - but this would not be accessible either if the device is lost. On the flip-side, the cloud hosting case requires significant user or developer friction, as a preventive measure for a somewhat rare scenario.

In the case of mobile apps, storing the data locally and persisting to the cloud can be very similar. iOS and Android both give app developers built-in cloud-synced storage to work with. The lift might not be so bad.

> For those truly fearful of this scenario and willing to tolerate some friction, they can set up a client (whether branded as a backup app or just a regular messaging app) on a secondary backup device.

Definitely an option, although there are large parts of the world where most users have a single device. It feels like this would only be an option as a temporary means of scope-cutting, but wouldn't work as a complete long-term solution.